File: robot/ros_ws/src/modules/asm_mighty/mighty/src/mighty/convert_odom_to_state.cpp
#include <mighty/convert_odom_to_state.hpp>

OdometryToStateNode::OdometryToStateNode() : Node("odometry_to_state_node") {
  // Subscriber for nav_msgs/Odometry
  odometry_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
      "odom", 10, std::bind(&OdometryToStateNode::callback, this, std::placeholders::_1));

  // Publisher for dynus_interfaces/State — depth 1: only the latest state
  // matters and a deeper queue lets back-pressure stack up under Zenoh.
  rclcpp::QoS state_qos(rclcpp::KeepLast(1));
  state_qos.reliable().durability_volatile();
  state_publisher_ = this->create_publisher<dynus_interfaces::msg::State>("state", state_qos);
}

void OdometryToStateNode::callback(const nav_msgs::msg::Odometry::SharedPtr odom_msg) {
  // Construct the State message
  auto state_msg = dynus_interfaces::msg::State();
  state_msg.header.stamp = odom_msg->header.stamp;
  state_msg.header.frame_id = odom_msg->header.frame_id;

  // Set position from Odometry
  state_msg.pos.x = odom_msg->pose.pose.position.x;
  state_msg.pos.y = odom_msg->pose.pose.position.y;
  state_msg.pos.z = odom_msg->pose.pose.position.z;

  // Set velocity from Odometry
  state_msg.vel.x = odom_msg->twist.twist.linear.x;
  state_msg.vel.y = odom_msg->twist.twist.linear.y;
  state_msg.vel.z = odom_msg->twist.twist.linear.z;

  // Set orientation from Odometry
  state_msg.quat.x = odom_msg->pose.pose.orientation.x;
  state_msg.quat.y = odom_msg->pose.pose.orientation.y;
  state_msg.quat.z = odom_msg->pose.pose.orientation.z;
  state_msg.quat.w = odom_msg->pose.pose.orientation.w;

  // Publish the State message
  state_publisher_->publish(state_msg);
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<OdometryToStateNode>();

  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: robot/ros_ws/src/modules/asm_mighty/mighty/scripts/convert_odom_to_state.py
#!/usr/bin/env python3

# /* ----------------------------------------------------------------------------
#  * Copyright 2025, Kota Kondo, Aerospace Controls Laboratory
#  * Massachusetts Institute of Technology
#  * All Rights Reserved
#  * Authors: Kota Kondo, et al.
#  * See LICENSE file for the license information
#  * -------------------------------------------------------------------------- */

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from std_msgs.msg import Header
from geometry_msgs.msg import Vector3, Quaternion
from dynus_interfaces.msg import State

class OdometryToStateNode(Node):
    def __init__(self):
        super().__init__('odometry_to_state_node')

        # Subscriber for nav_msgs/Odometry
        self.odometry_sub = self.create_subscription(
            Odometry, 'odom', self.callback, 10)

        # Publisher for dynus_interfaces/State
        self.state_publisher = self.create_publisher(State, 'state', 10)

    def callback(self, odom_msg):
        # Construct the State message
        state_msg = State()
        state_msg.header = Header()
        state_msg.header.stamp = odom_msg.header.stamp
        state_msg.header.frame_id = odom_msg.header.frame_id

        # Set position from Odometry
        state_msg.pos = Vector3(
            x=odom_msg.pose.pose.position.x,
            y=odom_msg.pose.pose.position.y,
            z=odom_msg.pose.pose.position.z
        )

        # Set velocity from Odometry
        state_msg.vel = Vector3(
            x=odom_msg.twist.twist.linear.x,
            y=odom_msg.twist.twist.linear.y,
            z=odom_msg.twist.twist.linear.z
        )

        # Set orientation from Odometry
        state_msg.quat = Quaternion(
            x=odom_msg.pose.pose.orientation.x,
            y=odom_msg.pose.pose.orientation.y,
            z=odom_msg.pose.pose.orientation.z,
            w=odom_msg.pose.pose.orientation.w
        )

        # Publish the State message
        self.state_publisher.publish(state_msg)
        self.get_logger().info(f'Published State: {state_msg}')

def main(args=None):
    rclpy.init(args=args)
    node = OdometryToStateNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
